csharp/src/Ice/EndpointHostResolver.cs

```diff
@@ -2,63 +2,116 @@
+            }
+
+            public void Join() => _thread.Join();
+
+            public void Run()
+            {
+                try
+                {
+                    _communicator.RunEndpointHostResolver();
+                }
+                catch (System.Exception ex)
+                {
+                    _communicator.Logger.error($"exception in endpoint host resolver thread {_thread.Name}:\n{ex}");
+                }
+            }
+
+            public void Start(ThreadPriority priority)
+            {
+                _thread.Priority = priority;
+                _thread.Start();
+            }
+        }
+
+        private class ResolveEntry
         {
-            _communicator = communicator;
-            _protocol = communicator.ProtocolSupport;
-            _preferIPv6 = communicator.PreferIPv6;
-            _thread = new HelperThread(this);
-            updateObserver();
-            _thread.Start(Util.stringToThreadPriority(communicator.GetProperty("Ice.ThreadPriority")));
+            internal IEndpointConnectors callback;
+            internal IPEndpoint endpoint;
+            internal string host;
+            internal Instrumentation.IObserver? observer;
+            internal int port;
+            internal EndpointSelectionType selType;
+
+            internal ResolveEntry(string host, int port, EndpointSelectionType selType, IPEndpoint endpoint, IEndpointConnectors callback)
+            {
+                this.host = host;
+                this.port = port;
+                this.selType = selType;
+                this.endpoint = endpoint;
+                this.callback = callback;
+            }
         }
 
-        public void resolve(string host, int port, Ice.EndpointSelectionType selType, IPEndpoint endpoint,
-                            IEndpointConnectors callback)
+        private bool _endpointHostResolverDestroyed;
+        private LinkedList<ResolveEntry> _enpointHostResolverQueue = new LinkedList<ResolveEntry>();
+        private Instrumentation.IThreadObserver? _endpointHostResolverObserver;
+        private HelperThread _endpointHostResolverThread;
+
+        public void Resolve(string host, int port, EndpointSelectionType selType, IPEndpoint endpoint,
+            IEndpointConnectors callback)
         {
             //
             // Try to get the addresses without DNS lookup. If this doesn't work, we queue a resolve
             // entry and the thread will take care of getting the endpoint addresses.
             //
-            INetworkProxy? networkProxy = _communicator.NetworkProxy;
+            INetworkProxy? networkProxy = NetworkProxy;
             if (networkProxy == null)
             {
                 try
                 {
-                    List<EndPoint> addrs = Network.getAddresses(host, port, _protocol, selType, _preferIPv6, false);
+                    List<EndPoint> addrs = Network.getAddresses(host, port, ProtocolSupport, selType, PreferIPv6, false);
```

Comment:
We should rename Protocol to Transport in a future refactoring, when protocol refers to "transport protocol" as opposed to the Ice protocol [version] in use.